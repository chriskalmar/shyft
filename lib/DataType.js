
import { passOrThrow } from './util';


class DataType {

  constructor ({ name, description }) {

    // validate settings
    passOrThrow(name, 'Missing data type name')
    passOrThrow(description, 'Missing data type description')

    // configure data type
    this.name = name
    this.description = description
  }


  toString() {
    return this.name
  }

}


export default DataType
