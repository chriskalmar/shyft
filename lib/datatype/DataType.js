
import { passOrThrow } from '../util';


class DataType {

  constructor (setup = {}) {

    const {
      name,
      description,
    } = setup

    passOrThrow(name, () => 'Missing data type name')
    passOrThrow(description, () => `Missing description for data type '${name}'`)

    this.name = name
    this.description = description
  }


  toString() {
    return this.name
  }

}


export default DataType
