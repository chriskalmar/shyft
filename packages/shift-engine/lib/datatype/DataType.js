
import {
  passOrThrow,
  isFunction,
} from '../util';


class DataType {

  constructor (setup = {}) {

    const {
      name,
      description,
      mock,
      validate,
      enforceRequired,
      defaultValue,
      enforceIndex,
    } = setup

    passOrThrow(name, () => 'Missing data type name')
    passOrThrow(description, () => `Missing description for data type '${name}'`)

    passOrThrow(
      isFunction(mock),
      () => `'Missing mock function for data type '${name}'`
    )

    if (validate) {
      passOrThrow(
        isFunction(validate),
        () => `'Invalid validate function for data type '${name}'`
      )

      this.validator = validate
    }

    if (defaultValue) {
      passOrThrow(
        isFunction(defaultValue),
        () => `'Invalid defaultValue function for data type '${name}'`
      )

      this.defaultValue = defaultValue
    }

    this.name = name
    this.description = description
    this.mock = mock

    if (enforceRequired) {
      this.enforceRequired = enforceRequired
    }

    if (enforceIndex) {
      this.enforceIndex = enforceIndex
    }
  }


  validate = (value, context) => {
    if (value && this.validator) {
      this.validator(value, context)
    }
  }


  toString() {
    return this.name
  }

}


export default DataType


export const isDataType = (obj) => {
  return (obj instanceof DataType)
}
