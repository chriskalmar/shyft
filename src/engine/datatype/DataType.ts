import { Entity, ShadowEntity, ViewEntity } from '../..';
import { passOrThrow, isFunction } from '../util';
import { ComplexDataType } from './ComplexDataType';

export type DataTypeSetup = {
  name?: string;
  description?: string;
  mock?: () => any;
  validate?: () => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;
};

export type DataTypeFunction = (params: {
  setup: DataTypeSetup;
  entity?: Entity | ShadowEntity | ViewEntity;
}) => DataType | ComplexDataType;

export class DataType {
  name: string;
  description: string;
  mock?: () => any;
  validator?: (value: any, context: any) => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;

  constructor(setup: DataTypeSetup = {} as DataTypeSetup) {
    const {
      name,
      description,
      mock,
      validate,
      enforceRequired,
      defaultValue,
      enforceIndex,
    } = setup;

    passOrThrow(name, () => 'Missing data type name');
    passOrThrow(
      description,
      () => `Missing description for data type '${name}'`,
    );

    passOrThrow(
      isFunction(mock),
      () => `'Missing mock function for data type '${name}'`,
    );

    if (validate) {
      passOrThrow(
        isFunction(validate),
        () => `'Invalid validate function for data type '${name}'`,
      );

      this.validator = validate;
    }

    if (defaultValue) {
      passOrThrow(
        isFunction(defaultValue),
        () => `'Invalid defaultValue function for data type '${name}'`,
      );

      this.defaultValue = defaultValue;
    }

    this.name = name;
    this.description = description;
    this.mock = mock;

    if (enforceRequired) {
      this.enforceRequired = enforceRequired;
    }

    if (enforceIndex) {
      this.enforceIndex = enforceIndex;
    }
  }

  validate = async (value: any, context: any): Promise<void> => {
    if (value && this.validator) {
      await this.validator(value, context);
    }
  };

  toString(): string {
    return this.name;
  }
}

export const isDataType = (obj: unknown): obj is DataType => {
  return obj instanceof DataType;
};
