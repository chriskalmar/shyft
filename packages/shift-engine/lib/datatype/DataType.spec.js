
import DataType, { isDataType } from './DataType';

import {
  passOrThrow,
} from '../util';

import {
  DataTypeBoolean,
  DataTypeString,
} from './dataTypes';


describe('DataType', () => {

  it('should have a name', () => {

    function fn() {
      new DataType() // eslint-disable-line no-new
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have a description', () => {

    function fn() {
      new DataType({ // eslint-disable-line no-new
        name: 'example'
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should provide a mock function', () => {

    function fn() {
      new DataType({ // eslint-disable-line no-new
        name: 'example',
        description: 'Just some description',
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should return it\'s name', () => {

    const dataType = new DataType({
      name: 'someDataTypeName',
      description: 'Just some description',
      mock () {},
    })

    expect(dataType.name).toBe('someDataTypeName');
    expect(String(dataType)).toBe('someDataTypeName');

  })


  it.skip('should have a unique name', () => {
  })


  describe('isDataType', () => {


    it('should recognize objects of type DataType', () => {

      function fn() {
        passOrThrow(
          isDataType(DataTypeBoolean) &&
          isDataType(DataTypeString),
          () => 'This error will never happen'
        )
      }

      expect(fn).not.toThrow()

    })


    it('should recognize non-DataType objects', () => {

      function fn() {
        passOrThrow(
          isDataType({}) ||
          isDataType(function test() {}) ||
          isDataType(Error),
          () => 'Not a DataType object'
        )
      }


      expect(fn).toThrowErrorMatchingSnapshot();

    })

  })


})
