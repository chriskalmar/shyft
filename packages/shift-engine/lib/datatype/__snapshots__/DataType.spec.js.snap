// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataType isDataType should recognize non-DataType objects 1`] = `"Not a DataType object"`;

exports[`DataType should have a description 1`] = `"Missing description for data type 'example'"`;

exports[`DataType should have a name 1`] = `"Missing data type name"`;

exports[`DataType should provide a mock function 1`] = `"'Missing mock function for data type 'example'"`;
